File: go/cmd/fixtures-gen-behavior/diff.go
// SPDX-License-Identifier: MIT

package main

import (
	"fmt"
	"path/filepath"
	"strings"

	vstar "hop.top/vstar"
	"hop.top/vstar/diff"
)

// componentDiffFixture is one entry of a <name>.diff.json file:
// the changes diff.OfCalendar reports for one paired component.
//
// UID is the identity a port keys on; it is lifted out of Path
// (which the reference renders as "VCALENDAR.VTODO[uid=x]") so the
// common case needs no path parsing. Path is carried too because a
// component with no UID is addressed positionally ("VALARM[#0]")
// and UID is then empty.
//
// Subs holds the same shape recursively for changed sub-components
// (a VALARM inside a VEVENT). Ops is always present, empty list and
// all, so an identical pair renders as `"ops": []` rather than
// null.
type componentDiffFixture struct {
	UID  string                 `json:"uid"`
	Path string                 `json:"path"`
	Ops  []opFixture            `json:"ops"`
	Subs []componentDiffFixture `json:"subs,omitempty"`
}

// opFixture is one property-level change.
//
// Before/After are nullable: an add has no before, a remove has no
// after. Params carry the property's parameters as an ordered list
// of name/value pairs — a parameter-only change (same value, new
// TZID) is a "change" op whose before and after values are equal
// and whose param lists differ.
//
// No message field: the reference's rendering of a diff is prose
// and not part of the contract.
type opFixture struct {
	Op           string         `json:"op"`
	Property     string         `json:"property"`
	Before       *string        `json:"before"`
	After        *string        `json:"after"`
	BeforeParams []paramFixture `json:"before_params,omitempty"`
	AfterParams  []paramFixture `json:"after_params,omitempty"`
}

// paramFixture is one property parameter.
type paramFixture struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

// paramTZID and tzMontreal build the parameter-only change case.
const (
	paramTZID  = "TZID"
	tzMontreal = "America/Montreal"
)

// diffCase is one generated diff fixture: two calendars and the op
// kinds the pair is authored to produce.
type diffCase struct {
	name string
	ops  []string
	a    vstar.Calendar
	b    vstar.Calendar
}

// writeDiffFamily writes <name>.a.ics, <name>.b.ics and
// <name>.diff.json per case and returns the file count.
func writeDiffFamily(dir string) (int, error) {
	n := 0
	for _, tc := range diffCases() {
		got := diff.OfCalendar(tc.a, tc.b)
		fixtures := componentDiffFixtures(got)
		if err := assertDiffOps(tc, fixtures); err != nil {
			return 0, err
		}
		if err := writeCalendar(filepath.Join(dir, tc.name+".a.ics"), tc.a); err != nil {
			return 0, err
		}
		if err := writeCalendar(filepath.Join(dir, tc.name+".b.ics"), tc.b); err != nil {
			return 0, err
		}
		if err := writeJSON(filepath.Join(dir, tc.name+".diff.json"), fixtures); err != nil {
			return 0, err
		}
		n += 3
	}
	return n, nil
}

// componentDiffFixtures projects the reference's diff onto the
// fixture shape. No sorting happens here: diff.OfCalendar already
// emits components in pairing order and properties sorted by name
// (case-insensitive), and that ordering is the documented contract
// a port reproduces.
func componentDiffFixtures(ds []diff.ComponentDiff) []componentDiffFixture {
	out := make([]componentDiffFixture, 0, len(ds))
	for _, d := range ds {
		subs := nonEmpty(d.SubDiffs)
		var nested []componentDiffFixture
		if len(subs) > 0 {
			nested = componentDiffFixtures(subs)
		}
		out = append(out, componentDiffFixture{
			UID:  uidFromPath(d.Path),
			Path: d.Path,
			Ops:  opFixtures(d.Properties),
			Subs: nested,
		})
	}
	return out
}

// nonEmpty drops sub-diffs that record no change. diff.OfCalendar
// filters these at the top level but carries them nested; a port
// comparing against the fixture should see only real changes.
//
// Returns nil, not an empty slice, when nothing survives: `subs`
// carries omitempty, so an empty slice would be dropped from the
// JSON and come back nil on decode. Producing nil directly keeps
// the fixture round-trippable — encode, decode, compare.
func nonEmpty(ds []diff.ComponentDiff) []diff.ComponentDiff {
	var out []diff.ComponentDiff
	for _, d := range ds {
		if !d.Empty() {
			out = append(out, d)
		}
	}
	return out
}

func opFixtures(pds []diff.PropertyDiff) []opFixture {
	out := make([]opFixture, 0, len(pds))
	for _, pd := range pds {
		f := opFixture{
			Op:       opToken(pd.Op),
			Property: propertyName(pd),
		}
		switch pd.Op {
		case diff.OpAdded:
			f.After = strPtr(pd.Property.Value)
			f.AfterParams = paramFixtures(pd.Property.Params)
		case diff.OpRemoved:
			f.Before = strPtr(pd.Property.Value)
			f.BeforeParams = paramFixtures(pd.Property.Params)
		case diff.OpChanged:
			f.Before = strPtr(pd.Old.Value)
			f.After = strPtr(pd.Property.Value)
			f.BeforeParams = paramFixtures(pd.Old.Params)
			f.AfterParams = paramFixtures(pd.Property.Params)
		}
		out = append(out, f)
	}
	return out
}

// propertyName reports the name the op is about. For a removal the
// b-side property is zero, so the name lives on Property for adds
// and changes and on Property for removals too (the reference puts
// the surviving side there in every case).
func propertyName(pd diff.PropertyDiff) string {
	if pd.Property.Name != "" {
		return pd.Property.Name
	}
	return pd.Old.Name
}

// The op tokens a <name>.diff.json records. The reference's
// DiffOp.String() is capitalized for human display; the fixtures use
// these lowercase forms so ports need no case convention of their
// own.
const (
	opAdd     = "add"
	opRemove  = "remove"
	opChange  = "change"
	opUnknown = "unknown"
)

// opToken renders a DiffOp as its lowercase fixture token.
func opToken(op diff.DiffOp) string {
	switch op {
	case diff.OpAdded:
		return opAdd
	case diff.OpRemoved:
		return opRemove
	case diff.OpChanged:
		return opChange
	default:
		return opUnknown
	}
}

func paramFixtures(params []vstar.Param) []paramFixture {
	if len(params) == 0 {
		return nil
	}
	out := make([]paramFixture, 0, len(params))
	for _, p := range params {
		out = append(out, paramFixture{Name: p.Name, Value: p.Value})
	}
	return out
}

func strPtr(s string) *string { return &s }

// uidFromPath lifts the uid out of a rendered diff path such as
// "VCALENDAR.VTODO[uid=todo-1]". Returns "" for a positional path
// ("VCALENDAR.VALARM[#0]") — the component has no UID to key on.
func uidFromPath(path string) string {
	const marker = "[uid="
	i := strings.LastIndex(path, marker)
	if i < 0 || !strings.HasSuffix(path, "]") {
		return ""
	}
	return path[i+len(marker) : len(path)-1]
}

// assertDiffOps verifies the case produced the op kinds it was
// authored for, so a refactor cannot quietly hollow out a fixture.
// An empty want list asserts the whole diff is empty.
func assertDiffOps(tc diffCase, fixtures []componentDiffFixture) error {
	seen := map[string]bool{}
	var walk func([]componentDiffFixture)
	walk = func(ds []componentDiffFixture) {
		for _, d := range ds {
			for _, o := range d.Ops {
				seen[o.Op] = true
			}
			walk(d.Subs)
		}
	}
	walk(fixtures)
	if len(tc.ops) == 0 {
		if len(fixtures) != 0 {
			return fmt.Errorf("diff case %q: expected no changes, got %d component diff(s)", tc.name, len(fixtures))
		}
		return nil
	}
	for _, want := range tc.ops {
		if !seen[want] {
			return fmt.Errorf("diff case %q: expected a %q op, got %v", tc.name, want, seen)
		}
	}
	return nil
}

// diffCases enumerates the generated pairs. Each isolates one
// change kind so a failing port sees exactly one thing wrong.
func diffCases() []diffCase {
	base := func(uid string, props ...vstar.Property) vstar.Component {
		return hashed(withProps(
			bare(vstar.CompTodo),
			append([]vstar.Property{
				prop("UID", uid),
				prop("DTSTAMP", stampTime),
				prop("SUMMARY", "Ship the fixtures"),
			}, props...)...,
		))
	}
	return []diffCase{
		{
			name: "property_added",
			ops:  []string{opAdd},
			a:    oneComponent("Diff-PropertyAdded-A", base("todo-add")),
			b: oneComponent("Diff-PropertyAdded-B",
				base("todo-add", prop("DUE", "20260101T000000Z"))),
		},
		{
			name: "property_removed",
			ops:  []string{opRemove},
			a: oneComponent("Diff-PropertyRemoved-A",
				base("todo-remove", prop("DUE", "20260101T000000Z"))),
			b: oneComponent("Diff-PropertyRemoved-B", base("todo-remove")),
		},
		{
			name: "value_changed",
			ops:  []string{opChange},
			a: oneComponent("Diff-ValueChanged-A",
				base("todo-change", prop("DUE", "20260101T000000Z"))),
			b: oneComponent("Diff-ValueChanged-B",
				base("todo-change", prop("DUE", "20260202T000000Z"))),
		},
		{
			name: "param_changed",
			ops:  []string{opChange},
			// Same value, different parameter: the change is
			// entirely in the params, which is why the fixture
			// format carries them separately from the value.
			a: oneComponent("Diff-ParamChanged-A",
				base("todo-param", vstar.Property{
					Name:   "DUE",
					Params: []vstar.Param{{Name: paramTZID, Value: tzMontreal}},
					Value:  "20260101T000000",
				})),
			b: oneComponent("Diff-ParamChanged-B",
				base("todo-param", vstar.Property{
					Name:   "DUE",
					Params: []vstar.Param{{Name: paramTZID, Value: "Europe/Paris"}},
					Value:  "20260101T000000",
				})),
		},
		{
			name: "component_added",
			ops:  []string{opAdd},
			a:    oneComponent("Diff-ComponentAdded-A", base("todo-keep")),
			b: twoComponents("Diff-ComponentAdded-B",
				base("todo-keep"),
				base("todo-new", prop("DUE", "20260301T000000Z"))),
		},
		{
			name: "component_removed",
			ops:  []string{opRemove},
			a: twoComponents("Diff-ComponentRemoved-A",
				base("todo-keep"),
				base("todo-gone", prop("DUE", "20260301T000000Z"))),
			b: oneComponent("Diff-ComponentRemoved-B", base("todo-keep")),
		},
		{
			name: "identical",
			ops:  nil,
			a:    oneComponent("Diff-Identical-A", base("todo-same", prop("DUE", "20260101T000000Z"))),
			b:    oneComponent("Diff-Identical-B", base("todo-same", prop("DUE", "20260101T000000Z"))),
		},
	}
}

// twoComponents wraps a pair of components in one calendar.
func twoComponents(slug string, a, b vstar.Component) vstar.Calendar {
	cal := oneComponent(slug, a)
	cal.Append(b)
	return cal
}

File: rs/tests/diff.rs
// SPDX-License-Identifier: MIT

//! The `behavior/diff` gate: every `<name>.a.ics` / `<name>.b.ics` pair
//! under `spec/behavior/diff/` and the `<name>.diff.json` the reference
//! records for it.
//!
//! **Order is the contract.** Components come out in pairing order, not
//! sorted by path; ops come out sorted by property name,
//! case-insensitively. The gate compares the two sequences positionally
//! so a port that sorts the component list fails here rather than
//! passing on a set comparison.

mod support;

use hop_top_vstar::codec::rfc5545;
use hop_top_vstar::diff::{
    calendar_equal, card_equal, component_equal, of_calendar, of_card, of_component, ComponentDiff,
    DiffOp, PropertyDiff,
};
use hop_top_vstar::{Calendar, Card, CompType, Component, Param, Property};
use support::json::Json;

/// The lowercase op tokens the fixtures record. The reference's
/// `DiffOp` `Display` is capitalized for humans; these are the wire
/// spellings.
fn op_token(op: DiffOp) -> &'static str {
    match op {
        DiffOp::Added => "add",
        DiffOp::Removed => "remove",
        DiffOp::Changed => "change",
    }
}

/// Lifts the uid out of a rendered path such as
/// `VCALENDAR.VTODO[uid=todo-1]`, mirroring the generator. Returns `""`
/// for a positional path — the component has no UID to key on.
fn uid_from_path(path: &str) -> &str {
    const MARKER: &str = "[uid=";
    match (path.rfind(MARKER), path.ends_with(']')) {
        (Some(i), true) => &path[i + MARKER.len()..path.len() - 1],
        _ => "",
    }
}

/// Reads a `<field>_params` list off a fixture op, defaulting to empty
/// when the key is absent (the generator omits it when there are none).
fn want_params(op: &Json, key: &str) -> Vec<(String, String)> {
    match op.get(key) {
        None | Some(Json::Null) => Vec::new(),
        Some(list) => list
            .array()
            .iter()
            .map(|p| {
                (
                    p.str_field("name").expect("param has a name").to_owned(),
                    p.str_field("value").expect("param has a value").to_owned(),
                )
            })
            .collect(),
    }
}

fn got_params(params: &[Param]) -> Vec<(String, String)> {
    params
        .iter()
        .map(|p| (p.name.clone(), p.value.clone()))
        .collect()
}

/// Reads a nullable string field: absent or `null` is `None`.
fn want_str<'a>(op: &'a Json, key: &str) -> Option<&'a str> {
    match op.get(key) {
        None | Some(Json::Null) => None,
        Some(Json::String(s)) => Some(s.as_str()),
        other => panic!("diff fixture: {key} is {other:?}, want string or null"),
    }
}

/// Asserts one `ops` list against one [`ComponentDiff`]'s properties.
fn assert_ops(got: &[PropertyDiff], want: &[Json], label: &str) {
    assert_eq!(
        got.len(),
        want.len(),
        "{label}: op count — got {:?}, want {:?}",
        got.iter()
            .map(|p| (op_token(p.op), p.property.name.as_str()))
            .collect::<Vec<_>>(),
        want.iter()
            .map(|o| (
                o.str_field("op").unwrap_or("?"),
                o.str_field("property").unwrap_or("?")
            ))
            .collect::<Vec<_>>()
    );

    for (i, (g, w)) in got.iter().zip(want).enumerate() {
        let at = format!("{label}.ops[{i}]");
        assert_eq!(
            op_token(g.op),
            w.str_field("op").expect("op token"),
            "{at}: op kind"
        );

        // The surviving side carries the name in every case; the
        // reference falls back to `old` only when `property` is zero.
        let name = if g.property.name.is_empty() {
            g.old.name.as_str()
        } else {
            g.property.name.as_str()
        };
        assert_eq!(
            name,
            w.str_field("property").expect("property"),
            "{at}: name"
        );

        let (before, before_params, after, after_params) = match g.op {
            DiffOp::Added => (
                None,
                Vec::new(),
                Some(g.property.value.as_str()),
                got_params(&g.property.params),
            ),
            DiffOp::Removed => (
                Some(g.property.value.as_str()),
                got_params(&g.property.params),
                None,
                Vec::new(),
            ),
            DiffOp::Changed => (
                Some(g.old.value.as_str()),
                got_params(&g.old.params),
                Some(g.property.value.as_str()),
                got_params(&g.property.params),
            ),
        };

        assert_eq!(before, want_str(w, "before"), "{at}: before");
        assert_eq!(after, want_str(w, "after"), "{at}: after");
        assert_eq!(
            before_params,
            want_params(w, "before_params"),
            "{at}: before_params"
        );
        assert_eq!(
            after_params,
            want_params(w, "after_params"),
            "{at}: after_params"
        );
    }
}

/// Asserts a `[ComponentDiff]` sequence against a fixture array,
/// positionally — the order IS the contract.
fn assert_component_diffs(got: &[ComponentDiff], want: &[Json], label: &str) {
    assert_eq!(
        got.len(),
        want.len(),
        "{label}: component count — got paths {:?}, want paths {:?}",
        got.iter().map(|d| d.path.as_str()).collect::<Vec<_>>(),
        want.iter()
            .map(|d| d.str_field("path").unwrap_or("?"))
            .collect::<Vec<_>>()
    );

    for (i, (g, w)) in got.iter().zip(want).enumerate() {
        let at = format!("{label}[{i}]");
        assert_eq!(
            g.path,
            w.str_field("path").expect("path"),
            "{at}: path (pairing order is the contract — do NOT sort)"
        );
        assert_eq!(
            uid_from_path(&g.path),
            w.str_field("uid").expect("uid"),
            "{at}: uid"
        );
        assert_ops(&g.properties, w.get("ops").expect("ops").array(), &at);

        // The fixture omits `subs` when empty; the reference carries
        // unchanged sub-diffs internally and the generator drops them.
        let nested: Vec<&ComponentDiff> = g.sub_diffs.iter().filter(|s| !s.is_empty()).collect();
        let want_subs: &[Json] = match w.get("subs") {
            None | Some(Json::Null) => &[],
            Some(list) => list.array(),
        };
        assert_eq!(
            nested.len(),
            want_subs.len(),
            "{at}: sub-diff count — got {:?}",
            nested.iter().map(|d| d.path.as_str()).collect::<Vec<_>>()
        );
        let owned: Vec<ComponentDiff> = nested.into_iter().cloned().collect();
        assert_component_diffs(&owned, want_subs, &format!("{at}.subs"));
    }
}

/// The gate proper: every `*.diff.json` under `spec/behavior/diff/`.
#[test]
fn diff_fixtures_match_the_reference() {
    let stems = support::behavior_stems("diff", ".diff.json");
    assert!(
        stems.len() >= 7,
        "behavior/diff shrank to {} cases",
        stems.len()
    );

    for stem in &stems {
        let a =
            rfc5545::parse(support::behavior_input("diff", &format!("{stem}.a.ics")).as_slice())
                .unwrap_or_else(|e| panic!("{stem}.a.ics: {e}"));
        let b =
            rfc5545::parse(support::behavior_input("diff", &format!("{stem}.b.ics")).as_slice())
                .unwrap_or_else(|e| panic!("{stem}.b.ics: {e}"));
        let want = support::behavior_json("diff", &format!("{stem}.diff.json"));

        assert_component_diffs(&of_calendar(&a, &b), want.array(), stem);
    }
}

/// `identical.diff.json` is `[]` — two equal documents produce no
/// entry, not an entry with no ops. This is the assertion that stops a
/// port from passing by reporting every component as changed.
#[test]
fn identical_calendars_produce_no_entries() {
    let a = rfc5545::parse(support::behavior_input("diff", "identical.a.ics").as_slice())
        .expect("identical.a.ics parses");
    let b = rfc5545::parse(support::behavior_input("diff", "identical.b.ics").as_slice())
        .expect("identical.b.ics parses");
    assert!(of_calendar(&a, &b).is_empty());
    // `calendar_equal` is NOT expected to agree here: the two fixtures
    // carry different PRODIDs by construction, and canonical bytes carry
    // PRODID while `of_calendar` deliberately ignores it. That the two
    // disagree is exactly the distinction both functions exist to draw.
    assert_ne!(a.prod_id, b.prod_id);
    assert!(!calendar_equal(&a, &b));
    assert!(calendar_equal(&a, &a));
}

/// Component order is **pairing order**, not sorted by path.
///
/// Every `spec/behavior/diff` fixture yields exactly one component
/// diff, so none of them can tell the two orders apart — a port that
/// sorted its output would pass the whole fixture gate. This case is
/// built so the two orders genuinely differ: `pair_subs` emits the
/// UID-bearing components first, in `(type, uid)` order, and only then
/// the UID-less ones positionally, so the UID-less VTODO lands *after*
/// a VTODO whose path sorts later than its own.
#[test]
fn component_order_is_pairing_order_not_sorted_by_path() {
    /// A VTODO with an optional UID and one `SUMMARY`, so every entry
    /// below has a real op and none is filtered out as unchanged.
    fn todo(uid: Option<&str>, summary: &str) -> Component {
        let mut c = Component::new(CompType::TODO);
        if let Some(uid) = uid {
            c.add(Property::new("UID", uid));
        }
        c.add(Property::new("SUMMARY", summary));
        c
    }

    let a = Calendar {
        prod_id: "-//Order//EN".into(),
        components: vec![
            todo(None, "positional before"),
            todo(Some("zzz-last"), "keyed before"),
        ],
    };
    let b = Calendar {
        prod_id: "-//Order//EN".into(),
        components: vec![
            todo(None, "positional after"),
            todo(Some("zzz-last"), "keyed after"),
        ],
    };

    let paths: Vec<String> = of_calendar(&a, &b).into_iter().map(|d| d.path).collect();
    assert_eq!(
        paths,
        vec!["VCALENDAR.VTODO[uid=zzz-last]", "VCALENDAR.VTODO[#0]",],
        "components come out in pairing order — UID-bearing first, then \
         positional. Sorting by path would put VTODO[#0] first, and the \
         single-component behavior fixtures cannot catch that."
    );
}

/// `X-VSTAR-HASH` is excluded from both sides: a diff reports what
/// changed in the content, not the restamped hash that followed.
#[test]
fn hash_property_is_excluded_from_both_sides() {
    let mut a = Component::new(CompType::TODO);
    a.add(Property::new("UID", "todo-hash"));
    a.add(Property::new("X-VSTAR-HASH", "sha256:aaaa"));

    let mut b = Component::new(CompType::TODO);
    b.add(Property::new("UID", "todo-hash"));
    b.add(Property::new("X-VSTAR-HASH", "sha256:bbbb"));

    assert!(
        of_component(&a, &b).is_empty(),
        "a differing X-VSTAR-HASH must not surface as a change"
    );
}

/// Ops are sorted by property name case-insensitively, which is a
/// different order from the property order on the wire.
#[test]
fn ops_sort_by_property_name_case_insensitively() {
    let mut a = Component::new(CompType::TODO);
    a.add(Property::new("UID", "todo-sort"));

    let mut b = Component::new(CompType::TODO);
    b.add(Property::new("UID", "todo-sort"));
    b.add(Property::new("zeta", "z"));
    b.add(Property::new("ALPHA", "a"));
    b.add(Property::new("mid", "m"));

    let d = of_component(&a, &b);
    let names: Vec<&str> = d
        .properties
        .iter()
        .map(|p| p.property.name.as_str())
        .collect();
    assert_eq!(names, vec!["ALPHA", "mid", "zeta"]);
}

/// The three equality functions route through canonical bytes, so
/// property order and parameter order are irrelevant.
#[test]
fn equality_ignores_property_and_parameter_order() {
    let mut a = Component::new(CompType::TODO);
    a.add(Property::new("UID", "todo-eq"));
    a.add(Property::new("DTSTAMP", "20260504T120000Z"));
    a.add(
        Property::new("DUE", "20260101T000000")
            .with_param("TZID", "America/Montreal")
            .with_param("VALUE", "DATE-TIME"),
    );

    let mut b = Component::new(CompType::TODO);
    b.add(
        Property::new("DUE", "20260101T000000")
            .with_param("VALUE", "DATE-TIME")
            .with_param("TZID", "America/Montreal"),
    );
    b.add(Property::new("DTSTAMP", "20260504T120000Z"));
    b.add(Property::new("UID", "todo-eq"));

    assert!(component_equal(&a, &b));
    assert!(of_component(&a, &b).is_empty());

    let mut different = b.clone();
    different.set(Property::new("SUMMARY", "changed"));
    assert!(!component_equal(&a, &different));
}

/// `of_card` diffs a card's properties; cards have no sub-components so
/// `sub_diffs` is always empty and `path` is the empty rendering root.
#[test]
fn card_diff_reports_property_changes() {
    let mut a = Card {
        uid: "urn:uuid:1".into(),
        kind: None,
        props: Vec::new(),
    };
    a.add(Property::new("FN", "Jad Bitar"));

    let mut b = a.clone();
    b.set(Property::new("FN", "J. Bitar"));

    assert!(!card_equal(&a, &b));
    let d = of_card(&a, &b);
    assert_eq!(d.path, "");
    assert!(d.sub_diffs.is_empty());
    assert_eq!(d.properties.len(), 1);
    assert_eq!(d.properties[0].op, DiffOp::Changed);
    assert_eq!(d.properties[0].old.value, "Jad Bitar");
    assert_eq!(d.properties[0].property.value, "J. Bitar");
}

/// `Display for ComponentDiff` renders the unified-diff-ish block, and
/// an empty diff renders as the empty string.
#[test]
fn display_renders_a_unified_diff_block() {
    let mut a = Component::new(CompType::TODO);
    a.add(Property::new("UID", "todo-render"));
    a.add(Property::new("DUE", "20260101T000000Z"));

    let mut b = Component::new(CompType::TODO);
    b.add(Property::new("UID", "todo-render"));
    b.add(Property::new("DUE", "20260202T000000Z"));
    b.add(Property::new("SUMMARY", "new"));

    let rendered = of_component(&a, &b).to_string();
    assert_eq!(
        rendered,
        "--- \n~ DUE: 20260101T000000Z -> 20260202T000000Z\n+ SUMMARY:new\n"
    );

    assert_eq!(of_component(&a, &a).to_string(), "");
}

/// `is_empty` recurses: a component whose own properties are unchanged
/// but whose sub-component changed is NOT empty.
#[test]
fn is_empty_recurses_into_sub_diffs() {
    let mut alarm_a = Component::new(CompType::ALARM);
    alarm_a.add(Property::new("UID", "alarm-1"));
    alarm_a.add(Property::new("ACTION", "DISPLAY"));

    let mut a = Component::new(CompType::EVENT);
    a.add(Property::new("UID", "evt-1"));
    a.sub.push(alarm_a.clone());

    let mut b = a.clone();
    b.sub[0].set(Property::new("ACTION", "AUDIO"));

    let d = of_component(&a, &b);
    assert!(d.properties.is_empty(), "top level itself is unchanged");
    assert!(
        !d.is_empty(),
        "a changed sub-component makes the diff non-empty"
    );
    assert_eq!(d.sub_diffs.len(), 1);
    assert_eq!(d.sub_diffs[0].path, "VALARM[uid=alarm-1]");
}

/// `of_calendar` deliberately does NOT diff `PRODID`: the component set
/// is what V* equality cares about.
#[test]
fn of_calendar_ignores_prodid() {
    let mut c = Component::new(CompType::TODO);
    c.add(Property::new("UID", "todo-prodid"));
    c.add(Property::new("DTSTAMP", "20260504T120000Z"));

    let a = Calendar {
        prod_id: "-//A//EN".into(),
        components: vec![c.clone()],
    };
    let b = Calendar {
        prod_id: "-//B//EN".into(),
        components: vec![c],
    };

    assert!(of_calendar(&a, &b).is_empty());
    // `calendar_equal` is the counterpart that DOES notice — it routes
    // through canonical bytes, which carry PRODID.
    assert!(!calendar_equal(&a, &b));
}
